agent-control/src/secret_retriever/k8s/retrieve.rs

```diff
@@ -0,0 +1,138 @@
+use crate::agent_control::config::{K8sConfig, OpAMPClientConfig};
+use crate::secret_retriever::OpampSecretRetriever;
+use crate::secrets_provider::SecretsProvider;
+use crate::secrets_provider::k8s_secret::K8sSecretProvider;
+
+pub struct K8sSecretRetriever<P> {
+    provider: P,
+    config: K8sConfig,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("{0}")]
+pub struct K8sRetrieverError(String);
+
+impl<P> K8sSecretRetriever<P>
+where
+    P: SecretsProvider,
+{
+    pub fn new(provider: P, config: K8sConfig) -> Self {
+        Self { provider, config }
+    }
+}
+
+impl<P> OpampSecretRetriever for K8sSecretRetriever<P>
+where
+    P: SecretsProvider,
+{
+    type Error = K8sRetrieverError;
+
+    fn retrieve(&self, _opamp_config: &OpAMPClientConfig) -> Result<String, Self::Error> {
+        let secret_path = K8sSecretProvider::build_secret_path(
+            &self.config.namespace,
+            &self.config.auth_secret.secret_name,
+            &self.config.auth_secret.secret_key_name,
+        );
+
+        self.provider.get_secret(&secret_path).map_err(|e| {
+            K8sRetrieverError(format!(
+                "K8s getting secret from k8s: secret: {secret_path}, error: {e}"
+            ))
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::agent_control::config::AuthSecret;
+    use std::fmt;
+
+    struct MockProvider {
+        expected_path: String,
+        should_fail: bool,
+        return_value: String,
+    }
+
+    #[derive(Debug)]
+    struct MockError(String);
+    impl fmt::Display for MockError {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "{}", self.0)
+        }
+    }
+    impl std::error::Error for MockError {}
+
+    impl SecretsProvider for MockProvider {
```

Comment:
Any particular reason to use a "custom mock" instead of [mockall::mock!](https://docs.rs/mockall/latest/mockall/macro.mock.html) here?